File: statistics.js
/**
 * PixelAgent City — Statistics Dashboard
 * Tracks and displays game economy, agent performance, and mini-game stats.
 */

class StatsDashboard {
    constructor() {
        this.history = {
            dailyIncome: [],    // { day, income, expense, net }
            agentPerformance: [], // { day, tasks, avgMood, avgEnergy }
            contractHistory: [], // { day, completed, failed }
        };
        this.maxHistory = 30;
    }

    // Record daily data (call on each day end)
    recordDay(game, manager) {
        try {
            const agents = manager ? Array.from(manager.agents.values()) : [];
            const day = game.day;

            // Economy
            this.history.dailyIncome.push({
                day,
                income: game.totalEarned,
                expense: game.totalSpent,
                net: game.totalEarned - game.totalSpent,
                coins: game.coins,
            });

            // Agent performance
            const avgMood = agents.length > 0
                ? agents.reduce((s, a) => s + (a.mood || 70), 0) / agents.length
                : 0;
            const avgEnergy = agents.length > 0
                ? agents.reduce((s, a) => s + (a.energy || 80), 0) / agents.length
                : 0;
            this.history.agentPerformance.push({
                day,
                count: agents.length,
                tasks: manager?.stats?.totalTasksCompleted || 0,
                avgMood: Math.round(avgMood),
                avgEnergy: Math.round(avgEnergy),
            });

            // Contracts
            this.history.contractHistory.push({
                day,
                completed: game.completedContracts,
                failed: game.failedContracts,
            });

            // Trim history
            if (this.history.dailyIncome.length > this.maxHistory) this.history.dailyIncome.shift();
            if (this.history.agentPerformance.length > this.maxHistory) this.history.agentPerformance.shift();
            if (this.history.contractHistory.length > this.maxHistory) this.history.contractHistory.shift();
        } catch (e) {
            if (window.PAC) PAC.ErrorHandler.log('stats', `recordDay failed: ${e.message}`);
        }
    }

    // Get summary stats
    getSummary(game, manager) {
        const agents = manager ? Array.from(manager.agents.values()) : [];
        const stats = manager?.stats || {};

        return {
            // Economy
            coins: game.coins,
            totalEarned: game.totalEarned,
            totalSpent: game.totalSpent,
            netProfit: game.totalEarned - game.totalSpent,
            dailyBurn: this._calcDailyBurn(game, agents),

            // Company
            day: game.day,
            level: game.companyLevel,
            reputation: game.reputation,
            completedContracts: game.completedContracts,
            failedContracts: game.failedContracts,
            successRate: game.completedContracts > 0
                ? ((game.completedContracts / (game.completedContracts + game.failedContracts)) * 100).toFixed(0)
                : '0',

            // Agents
            agentCount: agents.length,
            totalTasks: stats.totalTasksCompleted || 0,
            totalLines: stats.totalLinesWritten || 0,
            totalCommits: stats.totalCommits || 0,
            avgMood: agents.length > 0
                ? Math.round(agents.reduce((s, a) => s + (a.mood || 70), 0) / agents.length)
                : 0,
            avgEnergy: agents.length > 0
                ? Math.round(agents.reduce((s, a) => s + (a.energy || 80), 0) / agents.length)
                : 0,

            // Role distribution
            roleDistribution: this._getRoleDistribution(agents),

            // Performance trend
            trend: this._calcTrend(),
        };
    }

    _calcDailyBurn(game, agents) {
        let salary = 0;
        agents.forEach(a => {
            salary += game.salaries[a.role] || 15;
        });
        return salary;
    }

    _getRoleDistribution(agents) {
        const dist = {};
        agents.forEach(a => {
            dist[a.role] = (dist[a.role] || 0) + 1;
        });
        return dist;
    }

    _calcTrend() {
        const inc = this.history.dailyIncome;
        if (inc.length < 3) return 'neutral';
        const recent = inc.slice(-3);
        const avgRecent = recent.reduce((s, d) => s + d.net, 0) / recent.length;
        const older = inc.slice(-6, -3);
        if (older.length === 0) return 'neutral';
        const avgOlder = older.reduce((s, d) => s + d.net, 0) / older.length;
        if (avgRecent > avgOlder * 1.1) return 'up';
        if (avgRecent < avgOlder * 0.9) return 'down';
        return 'neutral';
    }

    // Render to a canvas (mini chart)
    drawMiniChart(canvas, type = 'coins') {
        if (!canvas) return;
        const ctx = canvas.getContext('2d');
        const W = canvas.width, H = canvas.height;
        ctx.clearRect(0, 0, W, H);

        // Background
        ctx.fillStyle = 'rgba(10, 14, 26, 0.6)';
        ctx.fillRect(0, 0, W, H);

        const data = this.history.dailyIncome;
        if (data.length < 2) {
            ctx.fillStyle = 'rgba(78, 205, 196, 0.5)';
            ctx.font = '10px Inter, sans-serif';
            ctx.textAlign = 'center';
            ctx.fillText('Chưa đủ dữ liệu', W / 2, H / 2);
            return;
        }

        const values = data.map(d => type === 'coins' ? d.coins : d.net);
        const min = Math.min(...values);
        const max = Math.max(...values);
        const range = max - min || 1;
        const PAD = 5;

        // Draw line
        ctx.strokeStyle = type === 'coins' ? '#4ecdc4' : (values[values.length - 1] >= 0 ? '#00d4aa' : '#ff4757');
        ctx.lineWidth = 2;
        ctx.beginPath();
        values.forEach((v, i) => {
            const x = PAD + (i / (values.length - 1)) * (W - PAD * 2);
            const y = PAD + (1 - (v - min) / range) * (H - PAD * 2);
            if (i === 0) ctx.moveTo(x, y);
            else ctx.lineTo(x, y);
        });
        ctx.stroke();

        // Fill gradient under line
        const lastX = PAD + ((values.length - 1) / (values.length - 1)) * (W - PAD * 2);
        ctx.lineTo(lastX, H - PAD);
        ctx.lineTo(PAD, H - PAD);
        ctx.closePath();
        const grad = ctx.createLinearGradient(0, 0, 0, H);
        grad.addColorStop(0, type === 'coins' ? 'rgba(78, 205, 196, 0.3)' : 'rgba(0, 212, 170, 0.3)');
        grad.addColorStop(1, 'rgba(10, 14, 26, 0)');
        ctx.fillStyle = grad;
        ctx.fill();
    }

    // Save/load
    saveData() {
        return { ...this.history };
    }

    loadData(data) {
        if (data) {
            this.history = {
                dailyIncome: data.dailyIncome || [],
                agentPerformance: data.agentPerformance || [],
                contractHistory: data.contractHistory || [],
            };
        }
    }
}

window.StatsDashboard = StatsDashboard;

File: game_engine/statistics.py
"""
statistics.py — JS StatsDashboard Parity
=======================================
Complements analytics.py with the exact summary/history shape from statistics.js.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from .models import GameState, ROLE_CONFIG


class StatsDashboard:
    def __init__(self, state: Optional[GameState] = None):
        self.state = state
        if state is not None:
            if not getattr(state, "analytics_history", None):
                state.analytics_history = {"dailyIncome": [], "agentPerformance": [], "contractHistory": []}
            self.history = state.analytics_history
        else:
            self.history = {"dailyIncome": [], "agentPerformance": [], "contractHistory": []}
        self.max_history = 30

    def record_day(self, state: Optional[GameState] = None) -> None:
        game = state or self.state
        if game is None:
            return
        agents = list(game.agents.values())
        day = game.day
        self.history.setdefault("dailyIncome", []).append({
            "day": day,
            "income": game.total_earned,
            "expense": game.total_spent,
            "net": game.total_earned - game.total_spent,
            "coins": game.coins,
        })
        avg_mood = sum((a.mood or 70) for a in agents) / len(agents) if agents else 0
        avg_energy = sum((a.energy or 80) for a in agents) / len(agents) if agents else 0
        self.history.setdefault("agentPerformance", []).append({
            "day": day,
            "count": len(agents),
            "tasks": game.agent_stats.get("totalTasksCompleted", game.total_tasks_done),
            "avgMood": round(avg_mood),
            "avgEnergy": round(avg_energy),
        })
        self.history.setdefault("contractHistory", []).append({
            "day": day,
            "completed": game.contracts_completed,
            "failed": game.contracts_failed,
        })
        for key in ("dailyIncome", "agentPerformance", "contractHistory"):
            if len(self.history[key]) > self.max_history:
                self.history[key] = self.history[key][-self.max_history:]
        if game is not None:
            game.analytics_history = self.history

    def get_summary(self, state: Optional[GameState] = None) -> Dict[str, Any]:
        game = state or self.state
        if game is None:
            return {}
        agents = list(game.agents.values())
        return {
            "coins": game.coins,
            "totalEarned": game.total_earned,
            "totalSpent": game.total_spent,
            "netProfit": game.total_earned - game.total_spent,
            "dailyBurn": self._calc_daily_burn(game, agents),
            "day": game.day,
            "level": game.level,
            "reputation": game.reputation,
            "completedContracts": game.contracts_completed,
            "failedContracts": game.contracts_failed,
            "successRate": f"{round((game.contracts_completed / (game.contracts_completed + game.contracts_failed)) * 100):.0f}" if game.contracts_completed > 0 else "0",
            "agentCount": len(agents),
            "totalTasks": game.agent_stats.get("totalTasksCompleted", game.total_tasks_done),
            "totalLines": game.agent_stats.get("totalLinesWritten", game.agent_stats.get("total_lines_written", 0)),
            "totalCommits": game.agent_stats.get("totalCommits", game.agent_stats.get("total_commits", 0)),
            "avgMood": round(sum((a.mood or 70) for a in agents) / len(agents)) if agents else 0,
            "avgEnergy": round(sum((a.energy or 80) for a in agents) / len(agents)) if agents else 0,
            "roleDistribution": self._get_role_distribution(agents),
            "trend": self._calc_trend(),
        }

    def _calc_daily_burn(self, game: GameState, agents: List[Any]) -> int:
        salary = 0
        for agent in agents:
            salary += ROLE_CONFIG.get(agent.role, ROLE_CONFIG["coder"]).get("salary", 15)
        return salary

    def _get_role_distribution(self, agents: List[Any]) -> Dict[str, int]:
        dist: Dict[str, int] = {}
        for agent in agents:
            dist[agent.role] = dist.get(agent.role, 0) + 1
        return dist

    def _calc_trend(self) -> str:
        inc = self.history.get("dailyIncome", [])
        if len(inc) < 3:
            return "neutral"
        recent = inc[-3:]
        avg_recent = sum(d.get("net", 0) for d in recent) / len(recent)
        older = inc[-6:-3]
        if len(older) == 0:
            return "neutral"
        avg_older = sum(d.get("net", 0) for d in older) / len(older)
        if avg_recent > avg_older * 1.1:
            return "up"
        if avg_recent < avg_older * 0.9:
            return "down"
        return "neutral"

    def save_data(self) -> Dict[str, List[Dict[str, Any]]]:
        return {k: list(v) for k, v in self.history.items()}

    def load_data(self, data: Optional[Dict[str, List[Dict[str, Any]]]]) -> None:
        if data:
            self.history = {
                "dailyIncome": data.get("dailyIncome", []),
                "agentPerformance": data.get("agentPerformance", []),
                "contractHistory": data.get("contractHistory", []),
            }
            if self.state is not None:
                self.state.analytics_history = self.history
